### PR TITLE
Fix custom properties sets followed by comments in no-extra-semicolons

### PR DESCRIPTION
**Pull request by hudochenkov**
> Which issue, if any, is this issue related to?

#2395

> Is there anything in the PR that needs further explanation?

Just failing tests. Have no time to look into the problem yet.

### CODE REVIEW

**Comment by jeddy3** [comment]
Tests look good, thanks. Labelling as `revision needed`.

**Comment by hudochenkov** [comment]
I tried to find a solution, but failed.

**Comment by evilebottnawi** [comment]
@hudochenkov i am champion this rule, i can see this late, it is difficult rule :smile: 
How about blur all comments? Maybe it is help?

**Comment by evilebottnawi** [comment]
it is break my mind, we support custom properties sets which not even working draft, also `postcss` parses them very badly, i think there is always a solution we just need to more think. But i hope the next verision `postcss` help us and change `raws` for `custom-*` (and other stuff).

/cc @ai what we can hear about the future version and `custom` stuff?

**Comment by ai** [comment]
You talking about adding `;` to rules `raws` in this cases:

```css
--font-bodytext: {
    font-variant-ligatures: common-ligatures, contextual;
    font-variant-numeric: oldstyle-nums, proportional-nums;
    font-feature-settings: "kern", "liga", "clig", "calt", "onum", "pnum";
    font-kerning: normal;
};
```

I am waiting some spec stabilization and we can add it in 6.0

**Comment by evilebottnawi** [comment]
@ai also about sets
```css
:root { 
  --foo: { color: red };
  --foo-bar: { color: red };
}
```

**Comment by evilebottnawi** [comment]
We use `*.toString()` it is potential incorrect position in non standard syntax, need create issue and change all `*.toString()` to `*.source.input.css` in all rules before release `8.0.0`.

File: lib/rules/no-extra-semicolons/index.js [hunk 1]
```diff
@@ -54,16 +54,29 @@ const rule = function (actual) {
...
+          || (node.type === "comment" && next
+            && (isCustomPropertySet(next) && node.parent.index(next) > 0 || next.type === "comment"))
+        ) {
           allowedSemi = 1
         }
 
+        const prev = node.prev()
+
+        if (isCustomPropertySet(node) && node.parent.index(node) > 0 && prev && prev.type === "comment") {
+          rawBeforeNode = prev.toString() + rawBeforeNode
```
Comment (hudochenkov) [inline lib/rules/no-extra-semicolons/index.js]: Should it be `prev.source.input.css`? :)

File: lib/rules/no-extra-semicolons/index.js [hunk 2]
```diff
@@ -54,16 +54,29 @@ const rule = function (actual) {
...
 
+        const next = node.next()
+
         // Forbid semicolon before first custom properties sets
-        if (isCustomPropertySet(node) && node.parent.index(node) > 0) {
+        // Also ignore comment before custom properties sets
+        if ((isCustomPropertySet(node) && node.parent.index(node) > 0)
+          || (node.type === "comment" && next
+            && (isCustomPropertySet(next) && node.parent.index(next) > 0 || next.type === "comment"))
+        ) {
```
Comment (hudochenkov) [inline lib/rules/no-extra-semicolons/index.js]: What do you think about splitting this `if` statement into `else if` or more `if` statements?

File: lib/rules/no-extra-semicolons/index.js (hunk 1 above)
Comment (evilebottnawi) [inline lib/rules/no-extra-semicolons/index.js]: @hudochenkov i do this in another PR :smile: https://github.com/stylelint/stylelint/pull/2396#issuecomment-283448628

File: lib/rules/no-extra-semicolons/index.js (hunk 2 above)
Comment (evilebottnawi) [inline lib/rules/no-extra-semicolons/index.js]: @hudochenkov it's possible, but why :smile:

Comment (hudochenkov) [inline lib/rules/no-extra-semicolons/index.js]: Too complicated. Hard to understand in current form.

File: lib/rules/no-extra-semicolons/index.js [hunk 3]
```diff
@@ -60,7 +60,7 @@ const rule = function (actual) {
         let allowedSemi = 0
 
         // Forbid semicolon before first custom properties sets
-        if ((isCustomPropertySet(node) && node.parent.index(node) > 0)) {
+        if (isCustomPropertySet(node) && node.parent.index(node) > 0) {
```
Comment (evilebottnawi) [inline lib/rules/no-extra-semicolons/index.js]: Thanks :smile:

**Comment by hudochenkov** [comment]
LGTM! Thanks!

**Comment by davidtheclark** [comment]
Added to Changelog:

> -   Fixed: `no-extra-semicolons` false positives for comments are custom property sets ([#2396](https://github.com/stylelint/stylelint/pull/2396)).
